[FLINK-28774][hive] Allow user to configure whether to enable sort not when it's for dynamic partition writing for HiveSource (#20419)
apache/flink: docs/content/docs/connectors/table/hive/hive_read_write.md

  </tbody>
</table>

> ### Configuration for Dynamic Partition Inserting 

Reviewer: ```suggestion
### Dynamic Partition Writing 
```

Reviewer: Introduce what's is dynamic partition writing at the beginning.

> By default, if it's for dynamic partition inserting, Flink will sort the data additionally by dynamic partition columns before writing into sink table.

Reviewer: Add the following words:

That means the sink will receive all elements of one partition and then all elements of another partition. Elements of different partitions will not be mixed. This is helpful for Hive sink to reduce the number of partition writers and improve writing performance by writing one partition at a time. Otherwise, too many partition writers may cause the OutOfMemory exception.

To avoid the extra sorting, you can set job configuration `table.exec.hive.dynamic-grouping.enabled` (`true` by default) to `false`.

> But with such configuration, it'll throw OOM exception if there are too may dynamic partitions.

Reviewer: ```suggestion
But with such a configuration, it may throw OutOfMemory exception if there are too many dynamic partitions.
```

> 

Reviewer: Add some hints about how to tune the dynamic partitioning. For example, add `DISTRIBUTED BY <partition_fields>` for hash shuffling when data is not skewed. You can also manually add `SORTED BY <partition_fields>` to achieve the same purpose as `table.exec.hive.dynamic-grouping.enabled=true`.
